src/shims/unix/linux/epoll.rs

```diff
@@ -411,20 +411,27 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
         let mut num_of_events: i32 = 0;
         let mut array_iter = this.project_array_fields(&event)?;
 
-        while let Some((epoll_key, epoll_return)) = ready_list.pop_first() {
-            // If the file description is fully close, the entry for corresponding FdID in the
-            // global epoll event interest table would be empty.
-            if this.machine.epoll_interests.get_epoll_interest(epoll_key.0).is_some() {
-                // Return notification to the caller if the file description is not fully closed.
-                if let Some(des) = array_iter.next(this)? {
-                    this.write_int_fields_named(
-                        &[
-                            ("events", epoll_return.events.into()),
-                            ("u64", epoll_return.data.into()),
-                        ],
-                        &des.1,
-                    )?;
-                    num_of_events = num_of_events.checked_add(1).unwrap();
+        while let Some(des) = array_iter.next(this)?
+            && !ready_list.is_empty()
+        {
+            let mut entry_written = false;
```

Comment:
I don't think using mutable state is necessary here. That makes reasoning about the code harder. The `!is_empty` check above also adds extra complexity that should not be needed.

Instead, I suggest you add a function like `ready_list_next` that returns the next item from the ready list that someone is still interested in:
```rust
while let Some((epoll_key, epoll_return)) = ready_list.pop_first() {
  // Ensure we are still interested in this event. The FD might have been closed since
  // the event was generated, in which case we are not interested any more.
  if this.machine.epoll_interests.get_epoll_interest(epoll_key.0).is_some() {
    return Some((epoll_key, epoll_return));
  }
}
return None;
```
Then you just use that method here, the loop here should become fairly simple.

Reply:
Updated. By the way, if the ``for`` loop you meant is for ``while let Some(des) = array_iter.next(this)? ``, it might not be possible because [ArrayIterator](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_const_eval/interpret/eval_context/struct.InterpCx.html#method.project_array_fields), which is the type of ``array_iter``, doesn't implement ``IntoIterator``.

Reply:
Ah right this has to remain `while let`, good point.